File: app/src/main/java/ru/yourok/m3u8loader/utils/PlayIntent.java
package ru.yourok.m3u8loader.utils;

import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.Uri;

import java.io.File;
import java.util.List;

import ru.yourok.loader.Store;

/**
 * Created by yourok on 03.01.17.
 */

public class PlayIntent {
    public static void start(Context context, String filename, String title) {
        String player = Store.getPlayer(context);
        Intent intent;
        switch (player) {
            //Chooser
            case "0":
                intent = getChooser(filename, title);
                break;
            //Default
            default:
            case "1":
                intent = getDefaultPlayer(filename, title);
                break;
            //MX Player
            case "2":
                intent = getMXPlayer(context, filename, title);
                break;
            case "3":
                intent = getKodiPlayer(context, filename, title);
                break;
        }
        if (intent.resolveActivity(context.getPackageManager()) != null)
            context.startActivity(intent);
    }

    private static Intent getChooser(String filename, String title) {
        Intent chooser = Intent.createChooser(getDefaultPlayer(filename, title), "   ");
        return chooser;
    }

    private static Intent getDefaultPlayer(String filename, String title) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        intent.setDataAndType(Uri.fromFile(new File(filename)), "video/mp4");
        intent.putExtra("title", title);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        return intent;
    }

    private static Intent getMXPlayer(Context context, String filename, String title) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        String pkg = "";

        PackageManager pm = context.getPackageManager();
        List<ApplicationInfo> packages = pm.getInstalledApplications(PackageManager.GET_META_DATA);
        for (ApplicationInfo packageInfo : packages)
            if (packageInfo.packageName.equals("com.mxtech.videoplayer.pro")) {
                pkg = "com.mxtech.videoplayer.pro";
                break;
            }

        if (pkg.isEmpty())
            for (ApplicationInfo packageInfo : packages)
                if (packageInfo.packageName.equals("com.mxtech.videoplayer.ad")) {
                    pkg = "com.mxtech.videoplayer.ad";
                    break;
                }

        if (pkg.isEmpty())
            return getChooser(filename, title);

        intent.setDataAndType(Uri.fromFile(new File(filename)), "video/*");
        intent.putExtra("title", title);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.setPackage(pkg);
        return intent;
    }

    private static Intent getKodiPlayer(Context context, String filename, String title) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        String pkg = "";

        PackageManager pm = context.getPackageManager();
        List<ApplicationInfo> packages = pm.getInstalledApplications(PackageManager.GET_META_DATA);
        for (ApplicationInfo packageInfo : packages)
            if (packageInfo.packageName.equals("org.xbmc.kodi")) {
                pkg = "org.xbmc.kodi";
                break;
            }

        if (pkg.isEmpty())
            return getChooser(filename, title);

        intent.setDataAndType(Uri.fromFile(new File(filename)), "video/*");
        intent.putExtra("title", title);
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.setPackage(pkg);
        return intent;
    }
}

File: app/src/main/java/ru/yourok/m3u8loader/player/PlayIntent.kt
package ru.yourok.m3u8loader.player

import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.support.v4.content.FileProvider
import android.widget.Toast
import ru.yourok.dwl.settings.Preferences
import ru.yourok.dwl.storage.Document
import ru.yourok.m3u8loader.BuildConfig
import java.io.File


/**
 * Created by yourok on 20.11.17.
 */
class PlayIntent(val context: Context) {
    fun start(filename: String, title: String) {
        try {
            val player = Preferences.get("Player", 0) as Int
            Handler(Looper.getMainLooper()).post {
                val intent = when (player) {
                //Chooser
                    0 -> getChooser(filename, title)
                    1 -> getDefaultPlayer(filename, title)
                //MX Player
                    2 -> getMXPlayer(context, filename, title)
                    3 -> getKodiPlayer(context, filename, title)
                //Default
                    else -> getDefaultPlayer(filename, title)
                }

                val pm = context.getPackageManager()
                if (intent.resolveActivity(pm) != null) {
                    context.startActivity(intent)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
            Toast.makeText(context, "Error open player: " + e.message, Toast.LENGTH_SHORT)
        }
    }

    private fun getChooser(filename: String, title: String): Intent {
        return Intent.createChooser(getDefaultPlayer(filename, title), "   ")
    }

    private fun getDefaultPlayer(filename: String, title: String): Intent {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        var uri = getUri(filename)
        intent.setDataAndType(uri, "video/mp4")
        intent.putExtra("title", title)
        return intent
    }

    private fun getMXPlayer(context: Context, filename: String, title: String): Intent {
        val intent = Intent(Intent.ACTION_VIEW)
        var pkg = ""

        val pm = context.getPackageManager()
        val packages = pm.getInstalledApplications(PackageManager.GET_META_DATA)
        for (packageInfo in packages)
            if (packageInfo.packageName == "com.mxtech.videoplayer.pro") {
                pkg = "com.mxtech.videoplayer.pro"
                break
            }

        if (pkg.isEmpty())
            for (packageInfo in packages)
                if (packageInfo.packageName == "com.mxtech.videoplayer.ad") {
                    pkg = "com.mxtech.videoplayer.ad"
                    break
                }

        if (pkg.isEmpty())
            return getChooser(filename, title)

        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

        val uri = getUri(filename)
        intent.setDataAndType(uri, "video/mp4")
        intent.putExtra("title", title)
        intent.`package` = pkg
        return intent
    }

    private fun getKodiPlayer(context: Context, filename: String, title: String): Intent {
        val intent = Intent(Intent.ACTION_VIEW)
        var pkg = ""

        val pm = context.getPackageManager()
        val packages = pm.getInstalledApplications(PackageManager.GET_META_DATA)
        for (packageInfo in packages)
            if (packageInfo.packageName == "org.xbmc.kodi") {
                pkg = "org.xbmc.kodi"
                break
            }

        if (pkg.isEmpty())
            return getChooser(filename, title)

        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

        val uri = getUri(filename)
        intent.setDataAndType(uri, "video/mp4")
        intent.putExtra("title", title)
        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        intent.`package` = pkg
        return intent
    }

    private fun getUri(filename: String): Uri {
        var uri = FileProvider.getUriForFile(context, BuildConfig.APPLICATION_ID + ".provider", File(filename))
        if (!File(filename).canWrite())
            uri = Document.openFile(filename)?.uri
        return uri
    }
}